pyiceberg/io/pyarrow.py

```diff
@@ -906,6 +986,76 @@ def after_map_value(self, element: pa.Field) -> None:
         self._field_names.pop()
 
 
+class _ConvertToIcebergWithFreshIds(PreOrderPyArrowSchemaVisitor[Union[IcebergType, Schema]]):
```

Comment:
Ouch, I still think we don't want to duplicate this whole visitor. Maybe include @HonahX his suggestion that he mentioned earlier.

Reply:
:) I remember @HonahX 's suggestion, but I'm not sure if we were able to find a way to avoid code duplication. In either case, we need to create a new pre-order visitor that assigns IDs... currently, all visitors are handled post-order.

Please let me know if I missed anything we've discussed!

Reply:
Ah, I see. At least we could call the `primitive()` of the other visitor, so we don't have to duplicate the mapping of Iceberg to Arrow primitives, because that's the part that concerns me the most.

Reply:
Sounds great. Let me refactor that component :)

Reply:
Since now we put everything behind the `Catalog.create_table`, does the visitor still need to be pre-order? The `new_table_metadata` will re-assign ids via a pre-order visitor later in the create_table workflow
https://github.com/apache/iceberg-python/blob/8464d713d69e6e2f4fa6f629362a27c940a097e6/pyiceberg/table/metadata.py#L399-L409

Reply:
Here's my understanding so far (please let me know if I overlooked anything):

- new_table_metadata requires a Schema
- Right now, a Schema cannot be created without field_ids assigned
- assign_fresh_schema_ids / [_SetFreshIDs](https://github.com/apache/iceberg-python/blob/0f08806d4431d5d60998dac1bca5780b6d2e2785/pyiceberg/schema.py#L1221) requires a unique ID per field to freshly assign the IDs, so we can't use a hack like assigning -1 for all the IDs, and then relying on assign_fresh_schema_ids to yield the correct result

I think the alternative would be to update **_ConvertToIceberg** to generate some arbitrarily unique IDs for each field in post-order and then relying on new_table_metadata to correctly assign the IDs from last_update_id

Reply:
Is there a reason why we decided on this approach to assign_fresh_schema_ids, over using a simple pre-order traversal to assign IDs as we see the fields? Is this because we want to accommodate for potential duplicates?

Reply:
Thanks @syun64 for the summarization. I fully agree with your reasoning. I raised that question just to make sure we all agree that the minimum requirement for the new visitor here is to assign **arbitrarily unique IDs** for each field. Based on our discussions, there are the following options:

1. update `_ConvertToIceberg` to generate some arbitrarily unique IDs for each field in post-order
2. Add a `_ConvertToIcebergWithFreshIds` post-order visitor
3. Add a `_ConvertToIcebergWithFreshIds` pre-order visitor + `pre_order_visit_pyarrow`

Option 1 has almost no code duplication and requires slight changes only. 
Option 2 requires some code duplication, which can be reduced by re-using `primitive()`(the current implementation) or reduced further by implementing some sort of inheritance as https://github.com/apache/iceberg-python/pull/219#discussion_r1456922888. 
Option 3 requires some code duplication, which can be reduced by re-using `primitive()` (the current implementation)

Since the schema generated by option 3 will still have some difference in field ids assignment than that by `assign_fresh_schema_ids`, I am more inclined toward option 1 and 2. I think we might not need the `pre_order_visit_pyarrow` and `PreOrderPyArrowSchemaVisitor`.

For option 1 and 2, I have no strong preference. If there is no further concern about mixing this special case for table creation with the visitor implementaion to read table, option 1 is optimal. If we still think it may be better to separate the logic from `_ConvertToIceberg`, considering this [concern](https://github.com/apache/iceberg-python/pull/219#issuecomment-1859273052), then option 2 can be a compromise choice. WDYT?

Reply:
These are all good suggestions @HonahX . Thank you for taking the time to write out the detailed response 😄 .

I'm not too keen on assigning 'random' IDs in post-order in _ConvertToIceberg or in _ConvertToIcebergWithFreshIds because that'll make it more difficult to write out the correct 'expected' outcome for the test case of the visitor when it is assigning fresh and unique IDs. Although it is deterministic, I think it's a bit difficult to leave logic that exists only for intermediary purpose, that is difficult to understand.

> I also do not know the exact reason and context here. Similarly, Java side does something more than a simple pre-order traversal. Looks like the java implementation can be traced back to the very beginning of iceberg.

If there's no additional context, I'm taking the liberty to entertain the idea of refactoring **_SetFreshIDs** visitor by analyzing its existing functionality, which is just to:
1. assign IDs in pre-order
2. track old field_id to reserved_id mapping just to ensure that old identifier_fields are mapped to the new ones at the end

If functionality (2) is the only reason we have the requirement of the field_ids needing to be unique... how do you folks feel about the idea of refactoring **_SetFreshIDs** to remove this constraint by using a set to keep track of identifier_field_ids when we **_get_and_increment** (assign) new IDs, and instead assign all field_ids as '-1' using **_ConvertToIceberg**/**_ConvertToIcebergFreshIDs**? And then use the refactored **_SetFreshIDs** visitor to assign the IDs in order?

```
class _SetFreshIDs(PreOrderSchemaVisitor[IcebergType]):
    """Traverses the schema and assigns monotonically increasing ids."""

    identifier_field_ids: Set[int]

    def __init__(self, next_id_func: Optional[Callable[[], int]] = None) -> None:
        self.identifier_field_ids= set()
        counter = itertools.count(1)
        self.next_id_func = next_id_func if next_id_func is not None else lambda: next(counter)

    def _get_and_increment(self, current_id: int) -> int:
        new_id = self.next_id_func()

        if current_id in identifier_field_ids:
            identifier_field_ids.pop(current_id)
            identifier_field_ids.add(new_id)
        return new_id
```

Reply:
Thanks @syun64 for sharing the comprehensive analysis!

> assign IDs in pre-order

Just want to make sure we are on the same page, my understanding is that `_SetFreshIds` is a little different from a normal pre-order visitor because it [increment the counter for each field and then visit children](https://github.com/apache/iceberg/pull/5627#discussion_r956788070). This makes it also rely on `reserved_ids` to store new ids for fields on the same level. We want to maintain this behavior when refactoring the `_SetFreshIds`. We could probably do it like the [jave implementation](https://github.com/apache/iceberg/blob/c07f2aabc0a1d02f068ecf1514d2479c0fbdd3b0/api/src/main/java/org/apache/iceberg/types/AssignFreshIds.java#L86-L97)

> assign all field_ids as '-1' using _ConvertToIceberg/_ConvertToIcebergFreshIDs? And then use the refactored _SetFreshIDs visitor to assign the IDs in order?

Overall, this approach sounds reasonable to me if we can find an easy way to refactor the `_SetFreshIds`. @Fokko, I'd appreciate your perspective on refactoring `_SetFreshIds`. Do you any issues with this approach?

Reply:
Thank you for your response! I'll give this a try and see how it turns out

Reply:
I also noticed that the Python implementation differs from the Java side. I don't know what the historical reason for this is. I couldn't find anything on the original PR about why this was done that way https://github.com/apache/iceberg/pull/5627. I'm okay with aligning this with the Java implementation.

> Overall, this approach sounds reasonable to me if we can find an easy way to refactor the _SetFreshIds. @Fokko, I'd appreciate your perspective on refactoring _SetFreshIds. Do you see any issues with this approach?

That's an okay approach, as long as the visitor to do this is hidden inside the package. We should not expose setting `-1` field IDs to the outside.

What I like about the current implementation is that the visitor can be used on its own. Converting a field with all `-1` IDs doesn't provide much value on its own.

I would love to get this in with the 0.6.0 release to simplify the creation of tables.

Reply:
> I would love to get this in with the 0.6.0 release to simplify the creation of tables.

Sounds good 👍  I'll try my best to bring this review to a consensus in the next few days. We already have a working version, so we just need to agree on the final details 😄 

>That's an okay approach, as long as the visitor to do this is hidden inside the package. We should not expose setting -1 field IDs to the outside.
> 
> What I like about the current implementation is that the visitor can be used on its own. Converting a field with all -1 IDs doesn't provide much value on its own.

I'm 🆗 with both approach, but I have a preference for the '-1' approach... and here's my current thought.

Both approaches will work, and in both approaches, we are introducing an intermediate state of the Iceberg Schema that should never be exposed. But what I like about this approach is that we are creating an intermediate Iceberg Schema that is very visibly wrong, that we could use to check the validity of the schema before we allow the erroneous intermediate state to materialize a lasting result. An example could be to introduce a new sanity check before writing the individual files, or before making a schema update, that checks that no two IDs are the same, or that the IDs are greater than 0.

Hence, I think the fact that the intermediate Iceberg Schema is "seemingly" correct actually feels more like a downside to me...

I'm curious to hear if that argument resonates with you or the rest of the group. Hopefully it will be the last discussion item before we settle on the final approach 🎉🎈